lib/spack/spack/test/installer.py

```diff
@@ -480,10 +483,12 @@ def test_bootstrapping_compilers_with_different_names_from_spec(
 ):
     with spack.config.override("config:install_missing_compilers", True):
         with spack.concretize.disable_compiler_existence_check():
-            spec = spack.spec.Spec("trivial-install-test-package%oneapi@22.2.0").concretized()
+            spec = spack.spec.Spec("trivial-install-test-package%oneapi@=22.2.0").concretized()
             spec.package.do_install()
 
-            assert spack.spec.CompilerSpec("oneapi@22.2.0") in spack.compilers.all_compiler_specs()
+            assert (
+                spack.spec.CompilerSpec("oneapi@=22.2.0") in spack.compilers.all_compiler_specs()
```

Comment:
This is making me think again that the externally facing comparison should do something a bit more intuitive, while `VersionList` can still use an internal comparator with the nice range properties you want -- does that make sense?

I can see people getting bitten by this, both for `CompilerSpec` and `Spec`

Reply:
Hm, as far as I can see, this is only fixed by reintroducing an equality `@22.2.0 == @=22.2.0`, which is what the PR is ultimately fixing :p it could also be written in terms of `any(s.satisfies(...))`, but even then I would write `@=22.2.0` because the test is looking for a specific version

Reply:
> The gist of the issue is, if the rhs contains concrete things, an abstract lhs doesn't compare equal; it's common to accidentally write an abstract version in the lhs, and it returns false w/o error, so easy to miss.

Yes exactly

> I don't think it's a good idea to guess what the intention of the user was and choose an interpretation accordingly, since that leads to bugs.

I think I'm fine with this because we're talking about `dict.__contains__`, and I think that will only be used internally. The one that's would be an issue is if `Spec.__contains__` had this behavior, because it hasn't in the past.  It has always traversed the spec to see if anything in the DAG satisfied the argument.  It'll continue to work as it has before.

Reply:
> The gist of the issue is, if the rhs contains concrete things, an abstract lhs doesn't compare equal; it's common to accidentally write an abstract version in the lhs, and it returns false w/o error, so easy to miss.

Yes exactly

> I don't think it's a good idea to guess what the intention of the user was and choose an interpretation accordingly, since that leads to bugs.

I think I'm fine with this because we're talking about `dict.__contains__`, and I think that will only be used internally. The one that's would be an issue is if `Spec.__contains__` had this behavior, because it hasn't in the past.  It has always traversed the spec to see if anything in the DAG satisfied the argument.  It'll continue to work as it has before.